tinygrad/codegen/uops.py

```diff
@@ -294,31 +294,19 @@ def print(self):
 
   def graph_rewrite(self, sink:UOp, pm:PatternMatcher):
     # recursive rewrite
-    changed = getenv("UOPS_REWRITE", 1)
-    run_cnt = 0
-    while changed:
-      changed = 0
-      @functools.lru_cache
-      def rewrite(u:UOp) -> UOp:
-        nonlocal changed
-        recurse_cnt = 0
-        up = u
-        # locally recursively rewrite
-        while (rewritten := pm.rewrite(up)):
-          assert recurse_cnt < 100, f"recursive_rewrite looped {up} <--> {rewritten}"
-          up = rewritten
-          recurse_cnt += 1
-        changed += recurse_cnt
-        # NOTE: this changes UOp, so we have to delete caches
-        up.src = tuple(rewrite(x) for x in up.src)
-        if 'parents' in up.__dict__: delattr(up, 'parents')
-        if 'cmp_tuple' in up.__dict__: delattr(up, 'cmp_tuple')
-        # replace with cached nodes
-        return self.nodes.setdefault(up.tuple(), up)
-      sink = rewrite(sink)
-      run_cnt += 1
-      assert run_cnt < 100, "exceeded 100 rewrite loops!"
-    return sink
+    if not getenv("UOPS_REWRITE", 1): return sink
+    def rewrite(u:UOp) -> UOp:
+      if (up := u) in cache: return cache[up]
+      for _ in range(100):
+        if (rewritten := pm.rewrite(up)) is None:
+          up = UOp(up.op, up.dtype, tuple(rewrite(x) for x in up.src), up.arg)
```

Comment:
is recreating UOp a required change for your cache to work?
this diff is rewriting the retry loops, removing lru, recreating UOp instead of overriding,

Reply:
on the one side i think this is simpler faster and correct

but i still need to think about it deeper